guidance/library/_regex.py

```diff
@@ -1,123 +1,150 @@
-                if all([notz.match(x) is not None for x in all_chars]):
-                    return any_char_but(temp_vals)
-
-            # if b''.join(temp_vals).decode('utf8') in nots:
-            #     # print('Negating', temp_vals)
-            #     return select([byte_range(x[0], x[1]) for x in negative_byte_range(temp_vals)])
-        return select(vals)
-    if node.head.value == "Concatenation":
-        ret = ""
-        for x in node.sons:
-            ret += tree_to_grammar(x, nots)
-        return ret
-    if node.head.value == "Kleene Star":
-        return zero_or_more(tree_to_grammar(node.sons[0], nots))
-
-
-def simplify_tree(regex):
-    """Merges sequence of byte-based concats or unions into single nodes, to make the grammar more compact"""
-    if len(regex.sons) and regex.head.value in ["Concatenation", "Union"]:
-        regex = merge_nodes(regex, regex.head.value)
-    regex.sons = [simplify_tree(x) for x in regex.sons]
-    return regex
-
-
-def merge_nodes(regex, op):
-    """Merges node sequences of concats and unions.
-    op must be 'Concatenation' or 'Union'
-    """
-    assert op in ["Concatenation", "Union"]
-    current = regex
-    if len(current.sons) and current.head.value == op:
-        val = []
-        to_visit = [current]
-        while to_visit:
-            current = to_visit.pop()
-            leaves = [x for x in current.sons if len(x.sons) == 0]
-            ops = [x for x in current.sons if len(x.sons) and x.head.value == op]
-            if len(leaves) + len(ops) == len(current.sons):
-                val.extend([x.head.value for x in leaves])
-                to_visit += [
-                    x for x in current.sons if len(x.sons) and x.head.value == op
-                ]
-        # if all I have is leaves, I can group them
-        if all([len(x.sons) == 0 for x in current.sons]) and current.head.value == op:
-            # val.extend([x.head.value for x in current.sons])
-            if op == "Concatenation":
-                val = "".join(val)
-                new_node = FakeNode(val, [])
-            elif op == "Union":
-                new_node = FakeNode("Union", [FakeNode(x, []) for x in val])
-            return new_node
-    if len(val) > 1:
-        if op == "Concatenation":
-            # merge all chars into a string, return a concat between the string (as a leaf) and whatever right-side children exist
-            val = "".join(val)
-            new_node = FakeNode(op, [FakeNode(val, []), merge_nodes(current, op)])
-        elif op == "Union":
-            # Merge all select options into a left leaf of a select, whatever operation is left becomes the right child.
-            new_node = FakeNode(
-                op,
-                [
-                    FakeNode("Union", [FakeNode(x, []) for x in val]),
-                    merge_nodes(current, op),
-                ],
+
+class Transformer:
```

Comment:
Minor, but given the overload of the word "Transformer" in our neck of the woods :)... I think we should shoot for something a bit off the path (i.e. "Converter" and "convert")? Or even something more verbose (i.e. RegexPatternConverter), since we won't be instantiating this class directly in many places?

Reply:
Agreed. And it's definitely private, so I'm happy with a verbose name.